network/p2p/gossip/handler.go

```diff
@@ -112,8 +149,76 @@ func (h Handler[T]) AppRequest(_ context.Context, _ ids.NodeID, _ time.Time, req
 		Gossip: gossipBytes,
 	}
 
-	h.sentN.Add(float64(len(response.Gossip)))
-	h.sentBytes.Add(float64(responseSize))
+	h.pullGossipSentN.Add(float64(len(response.Gossip)))
+	h.pullGossipSentBytes.Add(float64(responseSize))
 
 	return proto.Marshal(response)
 }
+
+func (h Handler[T, U]) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) {
+	msg := &sdk.PushGossip{}
+	if err := proto.Unmarshal(gossipBytes, msg); err != nil {
+		h.log.Debug("failed to unmarshal gossip", zap.Error(err))
+		return
+	}
+
+	forward := make([][]byte, 0, len(msg.Gossip))
+	receivedBytes := 0
+	sentBytes := 0
+	for _, bytes := range msg.Gossip {
+		gossipable := U(new(T))
+		if err := gossipable.Unmarshal(bytes); err != nil {
+			h.log.Debug("failed to unmarshal gossip",
+				zap.Stringer("nodeID", nodeID),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		receivedBytes += len(bytes)
+
+		if _, ok := h.set.Get(gossipable.GetID()); ok {
+			continue
+		}
+
+		if err := h.set.Add(gossipable); err != nil {
```

Comment:
I think this should be handled by the set. Currently it's racy if we are handling multiple `AppGossip` messages at the same time with the same tx...

If we changed this I think we could revert the signature change on Set as well